helion/_compiler/inductor_lowering.py

```diff
@@ -793,128 +795,259 @@ def apply_dot_requirements(
+
+    # Validate inputs
+    is_fp8 = lhs_dtype in [torch.float8_e4m3fn, torch.float8_e5m2] and rhs_dtype in [
         torch.float8_e4m3fn,
         torch.float8_e5m2,
-    ]:
-        datatype = None  # Let Triton use native FP8 computation
-
-    lhs_size = lhs_node.meta["val"].size()  # pyright: ignore[reportAttributeAccessIssue,reportOptionalMemberAccess]
-    rhs_size = rhs_node.meta["val"].size()  # pyright: ignore[reportAttributeAccessIssue,reportOptionalMemberAccess]
-    # check to see if it is 3D and the highest dim is 1
-    reduce_dim = False
-    if len(lhs_size) == 3:
-        env = CompileEnvironment.current()
-        lhs_dim_idx = env.get_block_id(lhs_size[0])
-        rhs_dim_idx = env.get_block_id(rhs_size[0])
-        if lhs_dim_idx is not None and rhs_dim_idx is not None:
-            lhs_dim_val = env.block_sizes[lhs_dim_idx]
-            rhs_dim_val = env.block_sizes[rhs_dim_idx]
-            if (
-                lhs_dim_val.from_config(ctx.cg.device_function.config) == 1
-                and rhs_dim_val.from_config(ctx.cg.device_function.config) == 1
-            ):
-                reduce_dim = True
+    ]
 
-    if not reduce_dim:
-        if with_acc:
-            precision_arg = (
-                f", input_precision={datatype!r}" if datatype is not None else ""
-            )
-            return expr_from_string(
-                f"tl.dot(lhs, rhs, acc=acc{precision_arg})",
-                lhs=lhs,
-                rhs=rhs,
-                acc=acc,  # pyright: ignore[reportArgumentType]
+    if is_scaled_mm:
+        # _scaled_mm requires FP8 inputs
+        valid_fp8_dtypes = (torch.float8_e4m3fn, torch.float8_e5m2)
+        assert lhs_dtype in valid_fp8_dtypes, (
+            f"torch._scaled_mm requires FP8 inputs, got {lhs_dtype} for first argument"
+        )
+        assert rhs_dtype in valid_fp8_dtypes, (
+            f"torch._scaled_mm requires FP8 inputs, got {rhs_dtype} for second argument"
+        )
+        # Currently only support 2D tensors
+        if len(lhs_shape) != 2 or len(rhs_shape) != 2:
+            raise NotImplementedError(
+                "torch._scaled_mm currently only supports 2D tensors"
             )
-        # without accumulator
-        precision_arg = (
-            f", input_precision={datatype!r}" if datatype is not None else ""
+    elif is_fp8:
+        # Regular ops shouldn't use FP8 directly
+        raise AssertionError("Please use `torch._scaled_mm` for FP8 matmul operations")
+
+    # Determine computation strategy
+    needs_reshape = _needs_3d_reshape(ctx, lhs_shape, rhs_shape)
+
+    # Build the computation
+    if needs_reshape:
+        mm_result = _compute_reshaped_dot(
+            ctx, node, lhs_ast, rhs_ast, acc_ast, lhs_shape, rhs_shape, is_fp8, datatype
+        )
+    else:
+        mm_result = _compute_direct_dot(lhs_ast, rhs_ast, acc_ast, is_fp8, datatype)
+
+    # Apply scaling for _scaled_mm
+    if is_scaled_mm:
```

Comment:
Haven't look to much into the details but since we added mx b200 support to scaled_mm, we might want to figure out in this pr how we map this to https://triton-lang.org/main/python-api/generated/triton.language.dot_scaled.html when the shapes align w/ this

Reply:
Curious does `tl.dot_scaled` only allow scale tensor's dtype to be `uint8`? IIUC for current FP8 tensor, the tensor-wise scale is `float32` dtype, so maybe we can't directly lower to `tl.dot_scaled`.

For MX dtype support (that lowers to `tl.dot_scaled`), we can probably have a standalone follow-up PR for it.

Reply:
Ohh good call out, I thought we could use the same op for doing the deepseek blockwise scaling on hopper but that is likely just a regular * in matmul main loop